perspective.el

```diff
@@ -968,6 +990,249 @@ perspective beginning with the given letter."
   (setq persp-show-modestring t)
   (persp-update-modestring))
 
+;; Symbols namespaced by persp--state (internal) and persp-state (user
+;; functions) provide functionality which allows saving perspective state on
+;; disk, and loading it into another Emacs session.
+;;
+;; The relevant commands are persp-state-save and persp-state-load (aliased to
+;; persp-state-restore).
+;;
+;; The (on-disk) data structure looks like this:
+;;
+;; {
+;;   :files [...]
+;;   :frames [
+;;     {
+;;       :persps {
+;;         "persp1" {
+;;           :buffers [...]
+;;           :windows [...]
+;;         }
+;;       }
+;;       :order [...]
+;;     }
+;;   ]
+;; }
+
+(defstruct persp--state-complete
+  files
+  frames)
+
+(defstruct persp--state-frame
+  persps
+  order)
+
+(defstruct persp--state-single
+  buffers
+  windows)
+
+(defun persp--state-interesting-buffer-p (buffer)
+  (and (buffer-name buffer)
+       (not (string-match "^[[:space:]]*\\*" (buffer-name buffer)))
+       (or (buffer-file-name buffer)
+           (with-current-buffer buffer (equal major-mode 'dired-mode)))))
+
+(defun persp--state-file-data ()
+  (cl-loop for buffer in (buffer-list)
+        if (persp--state-interesting-buffer-p buffer)
+        collect (or (buffer-file-name buffer)
+                    (with-current-buffer buffer ; dired special case
+                      default-directory))))
+
+(defun persp--state-window-state-massage (entry persp valid-buffers)
+  "This is a primitive code walker. It removes references to
+potentially problematic buffers from the data structure created
+by window-state-get and replaces them with references to the
+perspective-specific *scratch* buffer. Buffers are considered
+'problematic' when they have no underlying file, or are otherwise
+transient."
+  (cond
+    ;; base case 1
+    ((not (consp entry))
+     entry)
+    ;; base case 2
+    ((atom (cdr entry))
+     entry)
+    ;; leaf: modify this
+    ((eq 'leaf (car entry))
+     (lexical-let ((leaf-props (cdr entry)))
```

Comment:
It would help future maintainers if you documented/drew the shapes of things handled by this.

Reply:
I expanded the docstring for the walker: https://github.com/gcv/perspective-el/commit/4a0fe25f037b453f0bf63f65aec20ec18ebd4cec